File: python3/test/verify-canonicalization.py
##############################################################################
#                                                                            #
#  Copyright 2006-2019 WebPKI.org (http://webpki.org).                       #
#                                                                            #
#  Licensed under the Apache License, Version 2.0 (the "License");           #
#  you may not use this file except in compliance with the License.          #
#  You may obtain a copy of the License at                                   #
#                                                                            #
#      https://www.apache.org/licenses/LICENSE-2.0                           #
#                                                                            #
#  Unless required by applicable law or agreed to in writing, software       #
#  distributed under the License is distributed on an "AS IS" BASIS,         #
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  #
#  See the License for the specific language governing permissions and       #
#  limitations under the License.                                            #
#                                                                            #
##############################################################################

# Test program for JCS

import sys
import codecs
import os

from json import loads

from org.webpki.json.Canonicalize import canonicalize
from org.webpki.json.Canonicalize import serialize

from collections import OrderedDict

# Our test program
if not len(sys.argv) in (1,2):
    print ('[JSON-in-file]')
    sys.exit(1)

def readFile(name):
  return codecs.open(name, "r", "utf-8").read()

def oneTurn(fileName):
    print("\nFile: " + fileName)
    jsonData = readFile(os.path.join(inputPath,fileName))
    obj = loads(jsonData, object_pairs_hook=OrderedDict)
    canres = canonicalize(obj)
    expected = readFile(os.path.join(outputPath,fileName)).encode()
    if canres == expected:
      result = "\nSuccess"
    else:
      result = "\nFAILURE\n"
    byteCount = 0
    next = False
    utf8InHex = ''
    for c in canres:
      if byteCount > 0 and byteCount % 32 == 0:
        utf8InHex += '\n'
        next = False
      byteCount += 1
      if next:
        utf8InHex += ' '
      next = True
      utf8InHex += '{0:02x}'.format(c)
    print(utf8InHex + result)

testData = os.path.join(os.path.split(os.path.split(os.path.dirname(os.path.abspath(__file__)))[0])[0],'testdata')
inputPath = os.path.join(testData, 'input')
outputPath = os.path.join(testData, 'output')
if len(sys.argv) == 1:
    for fileName in os.listdir(inputPath):
        oneTurn(fileName)
else:
    oneTurn(sys.argv[1])

File: go/test/verify-canonicalization.go
//
//  Copyright 2006-2019 WebPKI.org (http://webpki.org).
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//
 
// This program verifies the JSON canonicalizer using a test suite
// containing sample data and expected output

package main

import (
    "fmt"
    "io/ioutil"
    "runtime"
    "path/filepath"
    "bytes"
    "webpki.org/jsoncanonicalizer"
)

func check(e error) {
    if e != nil {
        panic(e)
    }
}

var testdata string

var failures = 0

func read(fileName string, directory string) []byte {
    data, err := ioutil.ReadFile(filepath.Join(filepath.Join(testdata, directory), fileName))
    check(err)
    return data
}

func verify(fileName string) {
    actual, err := jsoncanonicalizer.Transform(read(fileName, "input"))
    check(err)
    recycled, err2 := jsoncanonicalizer.Transform(actual)
    check(err2)
    expected := read(fileName, "output")
    var utf8InHex = "\nFile: " + fileName
    var byteCount = 0
    var next = false
    for _, b := range actual {
        if byteCount % 32 == 0 {
            utf8InHex = utf8InHex + "\n"
            next = false
        }
        byteCount++
        if next {
            utf8InHex = utf8InHex + " "
        }
        next = true
        utf8InHex = utf8InHex + fmt.Sprintf("%02x", b)
    }
    fmt.Println(utf8InHex + "\n")
    if !bytes.Equal(actual, expected) || !bytes.Equal(actual, recycled) {
        failures++
        fmt.Println("THE TEST ABOVE FAILED!");
    }
}
 
func main() {
    _, executable, _, _ := runtime.Caller(0)
    testdata = filepath.Join(filepath.Dir(filepath.Dir(filepath.Dir(executable))), "testdata")
    fmt.Println(testdata)
    files, err := ioutil.ReadDir(filepath.Join(testdata, "input"))
    check(err)
    for _, file := range files {
        verify(file.Name())
    }
    if failures == 0 {
        fmt.Println("All tests succeeded!\n")
    } else {
        fmt.Printf("\n****** ERRORS: %d *******\n", failures)
    }
}
